pkg/federatedtypes/service.go

```diff
@@ -107,6 +107,28 @@ func (a *FederatedServiceAdapter) ObjectForCluster(template, override pkgruntime
 	return service
 }
 
+func (a *FederatedServiceAdapter)ObjectForUpdateOp(desiredObj, clusterObj pkgruntime.Object) pkgruntime.Object {
+	desiredService := desiredObj.(*corev1.Service)
+	clusterService := clusterObj.(*corev1.Service)
+
+	// ClusterIP and NodePort are allocated to Service by cluster, so retain the same if any while updating
+	desiredService.Spec.ClusterIP = clusterService.Spec.ClusterIP
+	for _, cPort := range clusterService.Spec.Ports {
+		for i, fPort := range clusterService.Spec.Ports {
+			if fPort.Name == cPort.Name && fPort.Protocol == cPort.Protocol && fPort.Port == cPort.Port {
+				desiredService.Spec.Ports[i].NodePort = cPort.NodePort
+			}
+		}
+	}
+	// ExternalIPs are not managed by Kubernetes, so retain the same if any while updating
+	desiredService.Spec.ExternalIPs = clusterService.Spec.ExternalIPs
+
+	// Pass the same ResourceVersion as in the cluster object for update operation, otherwise operation will fail.
+	desiredService.ResourceVersion = clusterService.ResourceVersion
```

Comment:
Why do you think this necessary?  

The problem of an implicit resource version in federation v1 (because a kube resource was the template vs a template being a field in v2) should not be present here.  And while a user could provide a resource version, I would expect that field to be ignored when constructing the desired form of a resource for a given cluster.

Reply:
For ExternalIPs, lets deal with that later. (Essentially i was trying to bring what existed in federation v1. this was added by this pr https://github.com/kubernetes/kubernetes/pull/51894)

For Resource version, it seems the k8s is rejecting the update operation if it is empty for service type. For other types it seems to be ok, if ResourceVersion is empty. Here is the error log, if i don't retain the Resourceversion for update operation.
```
E0409 12:30:17.613205   21138 controller.go:598] Failed to execute updates for FederatedService "8b2e62c0-16a8-40d2-a0d2-bfaa59fc62ba/test-service-b8j25": [Failed to update FederatedService "8b2e62c0-16a8-40d2-a0d2-bfaa59fc62ba/test-service-b8j25" in cluster test-cluster-7nqlv: Service "test-service-b8j25" is invalid: metadata.resourceVersion: Invalid value: "": must be specified for an update Failed to update FederatedService "8b2e62c0-16a8-40d2-a0d2-bfaa59fc62ba/test-service-b8j25" in cluster test-cluster-5z8kw: Service "test-service-b8j25" is invalid: metadata.resourceVersion: Invalid value: "": must be specified for an update]
```

Reply:
Ok, very curious.  Thanks for the info.